src/slam_toolbox_common.cpp

```diff
@@ -400,10 +522,14 @@ bool SlamToolbox::updateMap()
 
   // publish map as current
   map_.map.header.stamp = scan_header.stamp;
-  sst_->publish(
-    std::move(std::make_unique<nav_msgs::msg::OccupancyGrid>(map_.map)));
-  sstm_->publish(
-    std::move(std::make_unique<nav_msgs::msg::MapMetaData>(map_.map.info)));
+  if (sst_->is_activated()) {
```

Comment:
This doesn't be required. This shouldn't be possible if not active if this was done right

Reply:
OK.  
Now threads ( = `publishTransformLoop` , `publishVisualizations` and `run` ) is running even in the inactive state, so `updateMap` func is called periodically.
I will update it so that threads are started in `on_activate` and threads are joined and clear in `on_deactivate`.

Reply:
I believe if the node is a lifecycle node, even the Publisher object is somehow converted behind the scenes to a lifecycle publisher anyway

Reply:
- [240b36da8d981d0807f84ca975db992236f3e598] (same commit as https://github.com/SteveMacenski/slam_toolbox/pull/561#discussion_r1062394486 )
  - I have updated it so that threads are started in on_activate and threads are joined/clear in on_deactivate. 
- [89bcbc867c3823b4078f478f4bca8337a05251b6]
  - I noticed that `get_current_state` is not thred safe, so I added net mutex to lock it.
  - This is a bug in humble, and fixed in rolling. ( I have commented the issue link in codes. )

---

> I believe if the node is a lifecycle node, even the Publisher object is somehow converted behind the scenes to a lifecycle publisher anyway

As far as I checked rclcpp codes and did a simple test, I think it's not a problem to upcast to the Publisher because it is a minimal wrapper that only gives active state management to the Publisher. However, I left it as the lifecyclePublisher just to be safe.

<details> <summary> reference code links </summary>

- LifecyclePublisher : inherit from ManagedEntity and Publisher
  - https://github.com/ros2/rclcpp/blob/humble/rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_publisher.hpp
- ManagerEntity : impl of active state
  - https://github.com/ros2/rclcpp/blob/humble/rclcpp_lifecycle/include/rclcpp_lifecycle/managed_entity.hpp
- LifecyclePublisher is only used in the `create_publisher` function. ( can be upcast to Publisher )
  - https://github.com/ros2/rclcpp/blob/humble/rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node_impl.hpp#L44-L59

</details>